### PR TITLE
Fix attribute differences in between refs and implementations

### PR DESCRIPTION
**Pull request by safern**
We noticed that a change in ApiCompat is causing ApiCompat to not check for difference in attributes. We're fixing that in: https://github.com/dotnet/arcade/pull/5862

This reacts to that fix.

### CODE REVIEW

File: src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs [hunk 1]
```diff
@@ -254,6 +254,7 @@ public partial class Socket : System.IDisposable
     {
         public Socket(System.Net.Sockets.SafeSocketHandle handle) { }
         public Socket(System.Net.Sockets.AddressFamily addressFamily, System.Net.Sockets.SocketType socketType, System.Net.Sockets.ProtocolType protocolType) { }
+        [System.Runtime.Versioning.MinimumOSPlatformAttribute("windows7.0")]
```
Comment (safern) [inline src/libraries/System.Net.Sockets/ref/System.Net.Sockets.cs:257]: @dotnet/ncl @adamsitnik this was missed in the PR that marked windows specific APIs.

File: src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs [hunk 2]
```diff
@@ -493,6 +493,7 @@ public static partial class Marshal
         [System.Runtime.Versioning.MinimumOSPlatformAttribute("windows7.0")]
         [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
         public static System.IntPtr CreateAggregatedObject(System.IntPtr pOuter, object o) { throw null; }
+        [System.Runtime.Versioning.MinimumOSPlatformAttribute("windows7.0")]
```
Comment (safern) [inline src/libraries/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs:496]: @adamsitnik @AaronRobinsonMSFT this one was missed as well.

File: src/libraries/System.Runtime/ref/System.Runtime.cs [hunk 3]
```diff
@@ -9971,12 +9971,12 @@ public sealed partial class FrameworkName : System.IEquatable<System.Runtime.Ver
         public static bool operator !=(System.Runtime.Versioning.FrameworkName? left, System.Runtime.Versioning.FrameworkName? right) { throw null; }
         public override string ToString() { throw null; }
     }
-    [System.AttributeUsageAttribute(System.AttributeTargets.Assembly | System.AttributeTargets.Class | System.AttributeTargets.Constructor | System.AttributeTargets.Event | System.AttributeTargets.Field | System.AttributeTargets.Method | System.AttributeTargets.Module | System.AttributeTargets.Property | System.AttributeTargets.Struct, AllowMultiple = true, Inherited = false)]
+    [System.AttributeUsageAttribute(System.AttributeTargets.Assembly | System.AttributeTargets.Class | System.AttributeTargets.Constructor | System.AttributeTargets.Enum | System.AttributeTargets.Event | System.AttributeTargets.Field | System.AttributeTargets.Method | System.AttributeTargets.Module | System.AttributeTargets.Property | System.AttributeTargets.Struct, AllowMultiple = true, Inherited = false)]
```
Comment (safern) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9974]: @adamsitnik @jeffhandley this is updating the attributes to match the implementation which allows to set them on enums as well.

**Review by AaronRobinsonMSFT — APPROVED** [review]
Interop looks good.

**Review by jeffhandley — APPROVED** [review]
Thanks, @safern. Sorry about the mismatches.

**Comment by safern** [comment]
> Sorry about the mismatches.

Oh that is hard to catch without the tooling. I just wanted to give you an FYI in case we needed to port this to P8 or just do that you know some APIs missed it in case someone reports that when trying the analyzer 😄

File: eng/DefaultGenApiDocIds.txt [hunk 4]
```diff
@@ -7,9 +7,11 @@ T:System.ComponentModel.EditorAttribute
 T:System.ComponentModel.ToolboxItemAttribute
 T:System.ComponentModel.TypeDescriptionProviderAttribute
 T:System.Configuration.ConfigurationPropertyAttribute
+T:System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute
```
Comment (MichalStrehovsky) [inline eng/DefaultGenApiDocIds.txt:10]: If the purpose of this file is `These attributes should be excluded from reference assemblies.`, then `DynamicallyAccessedMembersAttribute` and `RequiresUnreferencedCodeAttribute` shouldn't be added to this list. We need them in ref assemblies.

Comment (safern) [reply eng/DefaultGenApiDocIds.txt:10]: I thought the linker just depended on these being in the implementation. Thanks for the heads up, then I'll remove them and add them to the ref assemblies that need it.

Comment (MichalStrehovsky) [reply eng/DefaultGenApiDocIds.txt:10]: Linker does, but we have plans for a Roslyn analyzer that would consume the refs, so we've been trying to keep the refs annotated as well. Seems like some annotations slipped through.

**Review by jeffhandley — APPROVED** [review]

File: eng/DefaultGenApiDocIds.txt (hunk 4 above)
Comment (safern) [reply eng/DefaultGenApiDocIds.txt:10]: I see. So do we want to remove all the linker related attributes in sync, or just the ones I'm adding to this file as part of this PR?

Comment (safern) [reply eng/DefaultGenApiDocIds.txt:10]: I'll fix the ones that I'm seeing on this PR by reverting this change and the other ones with the tooling updated from Arcade we can do in a follow up PR.

Comment (safern) [reply eng/DefaultGenApiDocIds.txt]: Alright, I removed these 2 additions and fixed the missmatches.

Comment (MichalStrehovsky) [reply eng/DefaultGenApiDocIds.txt]: > So do we want to remove all the linker related attributes in sync, or just the ones I'm adding to this file as part of this PR?

Just the ones you're adding.

Which ones were you thinking of removing? DynamicDependencyAttribute and UnconditionalSuppressMessageAttribute don't have to be propagated to refs - they're implementation details that are relevant when analyzing the method bodies only and not interesting for dataflow analysis. Dataflow analysis only cares about DynamicallyAccessedMembers and RequiresUnreferencedCode.

> Alright, I removed these 2 additions and fixed the mismatches.

thank you!

Comment (safern) [reply eng/DefaultGenApiDocIds.txt]: Great! Thanks for elaborating!

**Review by Anipik — APPROVED** [review]

**Comment by safern** [comment]
FYI: @eerhardt

File: src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+Compat issues with assembly System.Text.Json:
+CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.DynamicallyAccessedMembersAttribute' exists on 'System.Text.Json.Serialization.JsonConverterAttribute.ConverterType' in the implementation but not the reference.
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt:2]: Why is this necessary? can't we add it to the ref?

Comment (safern) [reply src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt:2]: Good catch. I think I did it for downlevel tfm's but I'll double check.

I might have to do something similar to what I did in the Extensions project to import the attribute cs file into the ref to define them as internal since they're not available in downlevel tfms.

Comment (eerhardt) [reply src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt:2]: Yeah, we should be consistent here between Extensions and other downlevel projects.

I originally didn't put the attributes in the Extensions `ref` assembly because they are `internal`. But thinking more about this, if we ever get an analyzer for the linker we would need those attributes (even though they are internal) in the ref assemblies.

---
In reply to: [464612345](https://github.com/dotnet/runtime/pull/40185#discussion_r464612345) [](ancestors = 464612345)

File: src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt [hunk 6]
```diff
@@ -30,13 +30,9 @@ TypesMustExist : Type 'System.Runtime.InteropServices.WindowsRuntime.ReadOnlyArr
...
-TypesMustExist : Type 'System.Runtime.InteropServices.WindowsRuntime.ReturnValueNameAttribute' does not exist in the implementation but it does exist in the contract.
-TypesMustExist : Type 'System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeMarshal' does not exist in the implementation but it does exist in the contract.
-TypesMustExist : Type 'System.Runtime.InteropServices.WindowsRuntime.WriteOnlyArrayAttribute' does not exist in the implementation but it does exist in the contract.
-Total Issues: 36
+Compat issues with assembly System.Text.Json:
+CannotChangeAttribute : Attribute 'System.AttributeUsageAttribute' on 'System.Text.Json.Serialization.JsonConverterAttribute' changed from '[AttributeUsageAttribute(AttributeTargets.Class | AttributeTargets.Enum | AttributeTargets.Property | AttributeTargets.Struct, AllowMultiple=false)]' in the contract to '[AttributeUsageAttribute(AttributeTargets.Class | AttributeTargets.Enum | AttributeTargets.Field | AttributeTargets.Property | AttributeTargets.Struct, AllowMultiple=false)]' in the implementation.
```
Comment (ericstj) [inline src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:34]: These are all OK.  Attribute diffing is pretty rigid, it just errors whenever there is a difference.  Technically we could add handling for certain attributes and then make more lenient rules.  Here, the usage was broadened, which is compatible.

**Review by ericstj — APPROVED** [review]

File: src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt (hunk 5 above)
Comment (safern) [reply src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt:2]: Yeah. I added these cause I saw it was specifically excluded for non-net5.0 targets in the ref: 
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Text.Json/ref/System.Text.Json.cs#L521

However, it seems it was because these attributes are not available, so I'll update to what I did in Extensions.

Comment (safern) [reply src/libraries/System.Text.Json/src/MatchingRefApiCompatBaseline.txt]: Fixed. Thanks.

File: src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt (hunk 6 above)
Comment (safern) [reply src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:34]: Opened: https://github.com/dotnet/arcade/issues/5881

**Comment by safern** [comment]
Merging since failures are unrelated and pending pipelines are not affected by this change and I want to avoid merge conflicts or people adding attributes in other PRs and missing those and having to update this PR.

**Comment by adamsitnik** [comment]
@safern thanks for catching and fixing all the issues! It's great that you have improved the tooling as well!

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 7]
```diff
@@ -518,14 +518,9 @@ public abstract partial class JsonConverter
     public partial class JsonConverterAttribute : System.Text.Json.Serialization.JsonAttribute
     {
         protected JsonConverterAttribute() { }
-#if NETCOREAPP && !NETCOREAPP3_0
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:521]: **Question:** How come this was caught, but not the #if/def above 

```C#
public static partial class JsonSerializer {
#if NETCOREAPP && !NETCOREAPP3_0
...
```

Comment (safern) [reply src/libraries/System.Text.Json/ref/System.Text.Json.cs:521]: I think it was just my mistake when looking at the diff and I accidentally discarded that change and then gave it a thought but forgot to redo the removal.
